src/libcollections/slice.rs

```diff
@@ -1305,213 +1304,290 @@ impl<T: Clone> ToOwned for [T] {
 // Sorting
 ////////////////////////////////////////////////////////////////////////////////
 
-fn insertion_sort<T, F>(v: &mut [T], mut compare: F)
+/// Inserts `v[0]` into pre-sorted sequence `v[1..]` so that whole `v[..]` becomes sorted.
+///
+/// This is the integral subroutine of insertion sort.
+fn insert_head<T, F>(v: &mut [T], compare: &mut F)
     where F: FnMut(&T, &T) -> Ordering
 {
-    let len = v.len() as isize;
-    let buf_v = v.as_mut_ptr();
-
-    // 1 <= i < len;
-    for i in 1..len {
-        // j satisfies: 0 <= j <= i;
-        let mut j = i;
+    if v.len() >= 2 && compare(&v[0], &v[1]) == Greater {
         unsafe {
-            // `i` is in bounds.
-            let read_ptr = buf_v.offset(i) as *const T;
-
-            // find where to insert, we need to do strict <,
-            // rather than <=, to maintain stability.
-
-            // 0 <= j - 1 < len, so .offset(j - 1) is in bounds.
-            while j > 0 && compare(&*read_ptr, &*buf_v.offset(j - 1)) == Less {
-                j -= 1;
+            // A common method of doing insertion sort is swapping adjacent elements until the
+            // first one gets to it's destination. Swapping is slow whenever elements are big
+            // structures and therefore costly to copy.
+            //
+            // We opt for a different approach: copy the first element into a temporary variable,
+            // shift lesser elements succeeding it, and then copy it back into the remaining hole.
+            let mut tmp = NoDrop { value: ptr::read(&v[0]) };
+
+            // Panic safety:
+            //
+            // The guard holds the intermediate state of the insertion process. It knows where the
+            // hole currently is. If `compare` panics at any point during the process shifting, the
+            // guard will get dropped, copy the temporary variable into the hole, and thus ensure
+            // that `v` still holds every object it initially held exactly once.
+            let mut guard = Guard {
+                src: &mut tmp.value,
+                dst: &mut v[1],
+            };
+            ptr::copy_nonoverlapping(&v[1], &mut v[0], 1);
```

Comment:
Yes, it's slightly slower to do a single bulk `ptr::copy` after search. I'll explain that in a comment.